### PR TITLE
Move project from DerivedSample to DerivedBySample model.

### PR DESCRIPTION
**Pull request by UlysseFG**

### CODE REVIEW

File: backend/fms_core/migrations/0063_v4_10_0.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+ADMIN_USERNAME = 'biobankadmin'
+
+def move_project_fk_to_derivedbysample(apps, schema_editor):
+    DerivedBySample = apps.get_model("fms_core", "DerivedBySample")
+
+    with reversion.create_revision(manage_manually=True):
+        admin_user = User.objects.get(username=ADMIN_USERNAME)
+
+        reversion.set_comment("Copy to derived_by_sample model from derived_sample the project foreign key.")
```
Comment (nafiz1001) [inline backend/fms_core/migrations/0063_v4_10_0.py:15]: I would call rather call this move than copy ;)

File: backend/fms_core/models/sample.py [hunk 2]
```diff
@@ -140,8 +140,8 @@ def children(self) -> List["Sample"]:
     # Computed property for project relation
     @property
     def projects(self) -> List["Project"]:
-        #return [derived_sample.project_id for derived_sample in self.derived_samples] if self.id else []
-        queryset = self.derived_samples.filter(project__isnull=False).distinct("project")
+        #return [derived_by_sample.project_id for derived_by_sample in self.derived_by_samples] if self.id else []
```
Comment (nafiz1001) [inline backend/fms_core/models/sample.py:143]: why are you keeping this commented code 🤔

File: backend/fms_core/models/sample.py [hunk 3]
```diff
@@ -140,8 +140,8 @@ def children(self) -> List["Sample"]:
...
+        queryset = self.derived_by_samples.filter(project__isnull=False).distinct("project")
```
[the remark this answers is not in the document]
Comment (nafiz1001) [reply backend/fms_core/models/sample.py:144]: Oh right `related_name`

File: backend/fms_core/serializers.py [hunk 4]
```diff
@@ -452,7 +452,7 @@ class Meta:
                   'location_barcode', 'location_coord', 'container_full_location', 'site',
                   'current_volume', 'concentration', 'creation_date', 'collection_site', 'experimental_group',
                   'individual_name', 'sex', 'taxon', 'cohort', 'pedigree', 'father_name', 'mother_name',
-                  'quality_flag', 'quantity_flag', 'projects', 'depleted', 'is_library', 'derived_samples_counts', 'comment')
+                  'quality_flag', 'quantity_flag', 'project', 'depleted', 'is_library', 'derived_samples_counts', 'comment')
```
Comment (nafiz1001) [inline backend/fms_core/serializers.py:455]: I see that sample only has property `projects`. Are you sure this is correct?

Comment (nafiz1001) [inline backend/fms_core/serializers.py:455]: I see that sample has `projects` property but not `project`. So not sure if this is incorrect or not.

File: backend/fms_core/services/experiment_run_info.py [hunk 5]
```diff
@@ -185,7 +181,15 @@ def _generate_pooled_samples(experiment_run: ExperimentRun, pool: Sample) -> Lis
...
-        run_info_samples.append(run_info_sample)
+        project: Optional[Project] = derived_by_sample.project
+        if project is None:
+            raise Exception(f'Sample {derived_by_sample.sample} has no project.')
+        else:
+            run_info_sample.project_obj_id = project.id
+            run_info_sample.project_name = project.name
+            run_info_sample.external_project_id = project.external_id
+            run_info_sample.external_project_name = project.external_name
+            run_info_samples.append(run_info_sample)
```
Comment (nafiz1001) [inline backend/fms_core/services/experiment_run_info.py:192]: you actually adding type isn't convincing my editor that it is indeed an optional project ;)

Comment (nafiz1001) [reply backend/fms_core/services/experiment_run_info.py:192]: but do still keep it

File: backend/fms_core/template_importer/row_handlers/project_study_link_samples/project_study_link_samples.py [hunk 6]
```diff
@@ -59,10 +59,10 @@ def process_row_inner(self, sample, project, action):
             elif action['name'] == ADD_STUDY_ACTION:
                 # Make sure the sample is already associated to the project of the given study. In case of pool one of the samples has to be associated to the project
                 if sample_obj.is_pool:
-                    if not any([derived_sample.project == project_obj for derived_sample in sample_obj.derived_samples.all()]):
+                    if not any([derived_by_sample.project == project_obj for derived_by_sample in sample_obj.derived_by_samples.all()]):
                             self.errors['add_to_study'] = (f"One of the samples in pool [{sample_obj.name}] has to be associated to project [{project_obj.name}].")
                 else: 
-                    if sample_obj.derived_sample_not_pool.project != project_obj:
+                    if sample_obj.derived_by_samples.first().project != project_obj:
```
Comment (nafiz1001) [inline backend/fms_core/template_importer/row_handlers/project_study_link_samples/project_study_link_samples.py:65]: Why not check if all of the derived samples are associated to to the project?

File: backend/fms_core/viewsets/_constants.py [hunk 7]
```diff
@@ -176,7 +176,7 @@
     "status": CATEGORICAL_FILTERS,
     "external_id": CATEGORICAL_FILTERS,
     "external_name": CATEGORICAL_FILTERS,
-    **_prefix_keys("project_derived_samples__samples__", _sample_minimal_filterset_fields),
+    **_prefix_keys("project_derived_by_samples__sample__", _sample_minimal_filterset_fields),
```
Comment (nafiz1001) [inline backend/fms_core/viewsets/_constants.py:179]: Why does this in particular have only a single underscore? That feels off...

File: frontend/src/components/filters/descriptions.js [hunk 8]
```diff
@@ -648,7 +648,7 @@ export const POOLED_SAMPLES_FILTERS = {
     },
     project_name: {
         type: FILTER_TYPE.INPUT,
-        key: "derived_sample__project__name",
+        key: "project__name",
```
Comment (nafiz1001) [inline frontend/src/components/filters/descriptions.js:651]: I was wondering if this filtering made sense. Turns out, pooled-sample works with DerivedBySample so it's good.

File: backend/fms_core/models/sample.py (hunk 2 above)
Comment (UlysseFG) [reply backend/fms_core/models/sample.py:143]: Not sure really. I'll remove ...

File: backend/fms_core/services/experiment_run_info.py (hunk 5 above)
Comment (UlysseFG) [reply backend/fms_core/services/experiment_run_info.py:192]: those are hints not types ... they do not have any functional meaning ... your editor is right to remain doubtful.

File: backend/fms_core/services/experiment_run_info.py [hunk 9]
```diff
@@ -155,11 +155,7 @@ def _generate_run_info_samples(experiment_run: ExperimentRun) -> List[RunInfoSam
     samples = Sample.objects.filter(container=experiment_run.container)
 
     for sample in samples:
-        if (sample.is_pool):
-            generated_rows += _generate_pooled_samples(experiment_run, sample)
-        else:
-            run_info_sample = _generate_sample(experiment_run, sample, sample.derived_sample_not_pool)
-            generated_rows += [run_info_sample]
+        generated_rows += _generate_pooled_samples(experiment_run, sample)
```
Comment (nafiz1001) [inline backend/fms_core/services/experiment_run_info.py:158]: You sure you don't need to check if the sample is actually a pool?

File: backend/fms_core/viewsets/_constants.py (hunk 7 above)
Comment (UlysseFG) [reply backend/fms_core/viewsets/_constants.py:179]: This is the related_name for that model attribute ...

File: backend/fms_core/template_importer/row_handlers/project_study_link_samples/project_study_link_samples.py (hunk 6 above)
Comment (UlysseFG) [reply backend/fms_core/template_importer/row_handlers/project_study_link_samples/project_study_link_samples.py:65]: This is meant to be usable only on samples that are not pool ... a single sample only has one project and one derived_by_sample.

File: backend/fms_core/services/experiment_run_info.py (hunk 9 above)
Comment (UlysseFG) [reply backend/fms_core/services/experiment_run_info.py:158]: All samples are pools of one ...

File: backend/fms_core/services/sample.py [hunk 10]
```diff
@@ -243,14 +245,18 @@ def transfer_sample(process: Process,
 
             derived_samples_destination = sample_source.derived_samples.all()
             volume_ratios = {}
+            projects = {}
             for derived_sample in derived_samples_destination:
-                volume_ratios[derived_sample.id] = DerivedBySample.objects.get(sample=sample_source, derived_sample=derived_sample).volume_ratio
+                source_derived_by_sample = DerivedBySample.objects.get(sample=sample_source, derived_sample=derived_sample)
+                volume_ratios[derived_sample.id] = source_derived_by_sample.volume_ratio
+                projects[derived_sample.id] = source_derived_by_sample.project
```
Comment (nafiz1001) [inline backend/fms_core/services/sample.py:252]: you're getting `source_derived_by_sample` from `derived_samples_destination`?

Comment (nafiz1001) [reply backend/fms_core/services/sample.py:252]: Never mind, the sample is source, the derived sample is what the source and destination sample have in common.

File: backend/fms_core/template_importer/row_handlers/sample_submission/sample.py
Comment (nafiz1001) [inline backend/fms_core/template_importer/row_handlers/sample_submission/sample.py:1]: Is this part of this ticket (moving project field) also?

File: backend/fms_core/viewsets/_fetch_data.py [hunk 11]
```diff
@@ -362,7 +366,7 @@ def fetch_export_data(self, ids: List[int] =[]) -> Tuple[List, int]:
                 'pedigree': derived_sample["biosample__individual__pedigree"] if not is_pool or not is_library else None,
                 'quality_flag': ["Failed", "Passed"][sample["quality_flag"]] if sample["quality_flag"] is not None else None,
                 'quantity_flag': ["Failed", "Passed"][sample["quantity_flag"]] if sample["quantity_flag"] is not None else None,
-                'projects': derived_sample["project__name"] if not is_pool else None,
+                'project': projects[sample["first_project_id"]] if not is_pool else None,
```
Comment (nafiz1001) [inline backend/fms_core/viewsets/_fetch_data.py:369]: Ah. So this is where the singular project is coming from for the `SampleExportSerializer`. Right?

File: backend/fms_core/viewsets/_fetch_data.py [hunk 12]
```diff
@@ -394,7 +398,13 @@ def fetch_export_metadata(self, ids: List[int] =[]) -> Tuple[List, int]:
...
             'derived_samples'
         )
-        samples_by_derived = {s["derived_samples"]: s for s in self.queryset}
+        samples_by_derived = defaultdict(list)
+        for s in self.queryset:
+            if s["derived_samples"] is list:
+                for derived_sample in s["derived_samples"]:
+                    samples_by_derived[derived_sample].append(s)
+            else:
+                samples_by_derived[s["derived_samples"]].append(s)
```
Comment (nafiz1001) [inline backend/fms_core/viewsets/_fetch_data.py:407]: I wrote a helper function called `make_generator` for that reason ;)

Comment (nafiz1001) [inline backend/fms_core/viewsets/_fetch_data.py:407]: I wrote a helper function called make_generator for that reason ;)

File: backend/fms_core/viewsets/_fetch_data.py [hunk 13]
```diff
@@ -419,32 +440,32 @@ def fetch_export_metadata(self, ids: List[int] =[]) -> Tuple[List, int]:
...
+                biosample_id = derived_sample["biosample__id"]
+                metadata = metadata_per_biosample[biosample_id] if biosample_id in metadata_per_biosample else None
+                data = {
+                    'alias': derived_sample["biosample__alias"],
+                    'biosample_id': biosample_id ,
+                    'sample_name': sample["name"],
+                    'container_name': sample["container__name"],
+                    'container_barcode': sample["container__barcode"],
+                    'coordinates': sample["coordinate__name"],
+                    'project': derived_by_samples[derived_sample_id][sample["id"]],
```
Comment (nafiz1001) [inline backend/fms_core/viewsets/_fetch_data.py:463]: we need a better name for this dict than `derived_by_samples`. Maybe projectByDerivedBySampleIdAndSampleId?

**Review by nafiz1001 — APPROVED** [review]
My brain has melted. But I approve anyway.

**Pull request merged**